File: src/firefly/java/edu/caltech/ipac/firefly/visualize/draw/FootprintObj.java
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
package edu.caltech.ipac.firefly.visualize.draw;

import edu.caltech.ipac.firefly.visualize.ScreenPt;
import edu.caltech.ipac.firefly.visualize.ViewPortPt;
import edu.caltech.ipac.firefly.visualize.WebPlot;
import edu.caltech.ipac.util.dd.Region;
import edu.caltech.ipac.util.dd.RegionLines;
import edu.caltech.ipac.visualize.plot.Pt;
import edu.caltech.ipac.visualize.plot.WorldPt;

import java.util.ArrayList;
import java.util.List;

/**
 * Draw one or more closed polygons.
 *
 * This class can only be used with a WebPlot, You must use WorldPt
 *
 * @author tatianag, Trey
 * @version $Id: FootprintObj.java,v 1.14 2012/11/30 23:17:01 roby Exp $
 */
public class FootprintObj extends DrawObj {

	public static final int DEF_WIDTH = 1;

	public enum Style {
		STANDARD, HANDLED
	}

	private final List<WorldPt[]> _fpList;
	private Style _style = Style.STANDARD;
	private int lineWidth = DEF_WIDTH;

	/**
	 * pass a array of WorldPt that represents the corners of a Polygon
	 * 
	 * @param ptAry
	 *            the array of points of a polygon
	 *
	 */
	public FootprintObj(WorldPt[] ptAry) {
		super();
		_fpList = new ArrayList<WorldPt[]>(1);
		_fpList.add(ptAry);
	}

	/**
	 * pass a list of polygons, each polygon is an array of WorldPt that are the
	 * corners of the polygon
	 * 
	 * @param ptList
	 *            of list of arrays of points of a polygon
	 *
	 */
	public FootprintObj(List<WorldPt[]> ptList) {
		super();
		_fpList = ptList;
	}

	@Override
	public boolean getCanUsePathEnabledOptimization() {
		return _style == Style.STANDARD && lineWidth == 1;
	}

	@Override
	public int getLineWidth() {
		return DEF_WIDTH;
	}

	private void setLineWidth(int w) {
		lineWidth = w;
	}

	public List<WorldPt[]> getPos() {
		return _fpList;
	}

	public void setStyle(Style s) {
		_style = s;
	}

	public Style getStyle() {
		return _style;
	}

	// public double getScreenDistOLD(WebPlot plot, ScreenPt pt)
	// throws ProjectionException {
	// double minDist= Float.MAX_VALUE;
	//
	//
	// // TODO use distance to line instead of distance to point
	// for (WorldPt ptAry[] : getPos()) {
	// for (WorldPt wpt : ptAry) {
	// ScreenPt testPt= plot.getScreenCoords(wpt);
	// if (testPt != null) {
	// double dx= pt.getIX() - testPt.getIX();
	// double dy= pt.getIY() - testPt.getIY();
	// double dist= Math.sqrt(dx*dx + dy*dy);
	// if (dist > 0 && dist < minDist) {
	// minDist= dist;
	// }
	// }
	// }
	// }
	// return minDist;
	// }

	public double getScreenDist(WebPlot plot, ScreenPt pt) {
		double minDistSq = Double.MAX_VALUE;

		double distSq;
		for (WorldPt ptAry[] : getPos()) {
			ScreenPt last = null;
			int totX = 0;
			int totY = 0;
			for (WorldPt wpt : ptAry) {
				ScreenPt testPt = plot.getScreenCoords(wpt);
				if (last == null)
					last = plot.getScreenCoords(ptAry[ptAry.length - 1]);
				if (testPt != null) {
					distSq = ptSegDistSq(testPt.getIX(), testPt.getIY(), last.getIX(), last.getIY(), pt.getIX(),
							pt.getIY());
					totX += testPt.getIX();
					totY += testPt.getIY();
					if (distSq < minDistSq) {
						minDistSq = distSq;
					}
				}
				last = testPt;
			}
			float aveX = (float) totX / ptAry.length;
			float aveY = (float) totY / ptAry.length;
			distSq = distToPtSq(aveX, aveY, pt.getIX(), pt.getIY());
			if (distSq < minDistSq) {
				minDistSq = distSq;
			}
		}

		return Math.sqrt(minDistSq);
	}

	@Override
	public Pt getCenterPt() {
		double xSum = 0;
		double ySum = 0;
		double xTot = 0;
		double yTot = 0;

		for (WorldPt ptAry[] : getPos()) {
			for (WorldPt wpt : ptAry) {
				xSum += wpt.getX();
				ySum += wpt.getY();
				xTot++;
				yTot++;
			}
		}
		return new WorldPt(xSum / xTot, ySum / yTot);
	}

	public void draw(Graphics jg, WebPlot p, DrawingDef def, boolean useStateColor, boolean onlyAddToPath)
			throws UnsupportedOperationException {
		drawFootprint(jg, p, def, useStateColor, onlyAddToPath);
	}

	public void draw(Graphics g, DrawingDef def, boolean useStateColor, boolean onlyAddToPath)
			throws UnsupportedOperationException {
		throw new UnsupportedOperationException("this type only supports drawing with WebPlot");
	}

	private void drawFootprint(Graphics jg, WebPlot plot, DrawingDef def, boolean useStateColor,
			boolean onlyAddToPath) {

		boolean inView = false;
		for (WorldPt ptAry[] : _fpList) {
			for (WorldPt wpt : ptAry) {
				if (wpt != null && plot.pointInViewPort(wpt)) {
					inView = true;
					break;
				}
			}
			if (inView)
				break;
		}

		if (inView) {
			for (WorldPt ptAry[] : _fpList) {
				switch (_style) {
				case STANDARD:
					drawStandardFootprint(jg, ptAry, plot, def, useStateColor, onlyAddToPath);
					break;
				case HANDLED:
					drawHandledFootprint(jg, ptAry, plot, def, useStateColor);
					break;
				default:
					break;
				}
			}
		}
	}

	private void drawStandardFootprint(Graphics g, WorldPt[] ptAry, WebPlot plot, DrawingDef def, boolean useStateColor,
			boolean onlyAddToPath) {

		WorldPt wpt0 = ptAry[ptAry.length - 1];
		String color = calculateColor(def, useStateColor);
		if (!onlyAddToPath)
			g.beginPath(color, lineWidth);
		for (WorldPt wpt : ptAry) {
			ViewPortPt pt0 = plot.getViewPortCoords(wpt0);
			ViewPortPt pt = plot.getViewPortCoords(wpt);
			if (pt0 == null || pt == null)
				return;
			// g.drawLine(color, DEF_WIDTH, pt0.getIX(), pt0.getIY(),
			// pt.getIX(), pt.getIY());
			g.pathMoveTo(pt0.getIX(), pt0.getIY());
			if (!plot.coordsWrap(wpt0, wpt)) {
				g.pathLineTo(pt.getIX(), pt.getIY());
			}
			wpt0 = wpt;
		}
		if (!onlyAddToPath)
			g.drawPath();
	}

	private void drawHandledFootprint(Graphics jg, WorldPt[] ptAry, WebPlot plot, DrawingDef def,
			boolean useStateColor) {
		WorldPt wpt0 = ptAry[ptAry.length - 1];
		for (WorldPt wpt : ptAry) {
			if (!wpt.equals(wpt0)) {
				ViewPortPt pt0 = plot.getViewPortCoords(wpt0);
				ViewPortPt pt = plot.getViewPortCoords(wpt);
				if (pt0 == null || pt == null)
					return;
				wpt0 = wpt;
				String color = calculateColor(def, useStateColor);
				jg.drawLine(color, 1, pt0.getIX(), pt0.getIY(), pt.getIX(), pt.getIY());
				DrawUtil.drawHandledLine(jg, "red", pt0.getIX(), pt0.getIY(), pt.getIX(), pt.getIY());
			}
		}
	}

	@Override
	public boolean getSupportDuplicate() {
		return true;
	}

	@Override
	public DrawObj duplicate() {
		FootprintObj obj = new FootprintObj(_fpList);
		obj.setStyle(_style);
		obj.copySetting(this);
		return obj;
	}

	public static double distToPtSq(float x0, float y0, float x1, float y1) {
		double dx = x1 - x0;
		double dy = y1 - y0;
		return dx * dx + dy * dy;
	}

	@Override
	public List<Region> toRegion(WebPlot plot, DrawingDef def) {
		List<Region> retList = new ArrayList<Region>(_fpList.size());
		String color = calculateColor(def, false);
		for (WorldPt ptAry[] : _fpList) {
			RegionLines rl = new RegionLines(ptAry);
			rl.getOptions().setColor(color);
			retList.add(rl);
		}
		return retList;
	}

	/**
	 * Returns the square of the distance from a point to a line segment. The
	 * distance measured is the distance between the specified point and the
	 * closest point between the specified endpoints. If the specified point
	 * intersects the line segment in between the endpoints, this method returns
	 * 0.0.
	 *
	 * <br>
	 * COPIED from java.awt.geom.Line2D, all documentation of this method is
	 * from Line2D. I did change from doubles to int in the method since I did
	 * not need the double precision <br>
	 *
	 * @param X1,&nbsp;Y1
	 *            the coordinates of the beginning of the specified line segment
	 * @param X2,&nbsp;Y2
	 *            the coordinates of the end of the specified line segment
	 * @param PX,&nbsp;PY
	 *            the coordinates of the specified point being measured against
	 *            the specified line segment
	 * @return a double value that is the square of the distance from the
	 *         specified point to the specified line segment.
	 */
	public static double ptSegDistSq(int X1, int Y1, int X2, int Y2, int PX, int PY) {
		// Adjust vectors relative to X1,Y1
		// X2,Y2 becomes relative vector from X1,Y1 to end of segment
		X2 -= X1;
		Y2 -= Y1;
		// PX,PY becomes relative vector from X1,Y1 to test point
		PX -= X1;
		PY -= Y1;
		int dotprod = PX * X2 + PY * Y2;
		double projlenSq;
		if (dotprod <= 0) {
			// PX,PY is on the side of X1,Y1 away from X2,Y2
			// distance to segment is length of PX,PY vector
			// "length of its (clipped) projection" is now 0.0
			projlenSq = 0.0;
		} else {
			// switch to backwards vectors relative to X2,Y2
			// X2,Y2 are already the negative of X1,Y1=>X2,Y2
			// to get PX,PY to be the negative of PX,PY=>X2,Y2
			// the dot product of two negated vectors is the same
			// as the dot product of the two normal vectors
			PX = X2 - PX;
			PY = Y2 - PY;
			dotprod = PX * X2 + PY * Y2;
			if (dotprod <= 0) {
				// PX,PY is on the side of X2,Y2 away from X1,Y1
				// distance to segment is length of (backwards) PX,PY vector
				// "length of its (clipped) projection" is now 0.0
				projlenSq = 0.0;
			} else {
				// PX,PY is between X1,Y1 and X2,Y2
				// dotprod is the length of the PX,PY vector
				// projected on the X2,Y2=>X1,Y1 vector times the
				// length of the X2,Y2=>X1,Y1 vector
				projlenSq = (double) (dotprod * dotprod) / (double) (X2 * X2 + Y2 * Y2);
			}
		}
		// Distance to line is now the length of the relative point
		// vector minus the length of its projection onto the line
		// (which is zero if the projection falls outside the range
		// of the line segment).
		double lenSq = PX * PX + PY * PY - projlenSq;
		if (lenSq < 0) {
			lenSq = 0;
		}
		return lenSq;
	}

	/**
	 * Build world points translated to apt
	 * 
	 * @param plot
	 * @param apt
	 *            world point to translate to
	 */
	public void translateTo(WebPlot plot, WorldPt apt) {
		// Need to clear fist the original array of point...
		synchronized (_fpList) {
			ScreenPt pt = plot.getScreenCoords(apt);
			WorldPt[] arr = _fpList.get(0);
			WorldPt[] arr2 = new WorldPt[arr.length];

			_fpList.clear();
			int i = 0;
			for (WorldPt wpt : arr) {
				ScreenPt pti = plot.getScreenCoords(wpt);

				int x2 = pti.getIX();
				int y2 = pti.getIY();
				x2 += pt.getIX();
				y2 += pt.getIY();
				arr2[i++] = plot.getWorldCoords(new ScreenPt(x2, y2));
			}
			// ... and recreate the ones translated

			_fpList.add(arr2);
		}
	}

	/**
	 * Rotate all points around a center wc
	 * 
	 * @param plot
	 *            plot object
	 * @param angle
	 *            angle to rotate in radians
	 * @param wc
	 *            center world point to rotate around
	 */
	public void rotateAround(WebPlot plot, double angle, WorldPt wc) {
		ScreenPt center = plot.getScreenCoords(wc);
		synchronized (_fpList) {

			WorldPt[] arr = _fpList.get(0);
			WorldPt[] arr2 = new WorldPt[arr.length];

			_fpList.clear();
			int i = 0;
			for (WorldPt p1 : arr) {
				// TRANSLATE TO ORIGIN
				ScreenPt pti = plot.getScreenCoords(p1);
				double xc = center.getX();
				double x1 = pti.getX() - xc;
				double yc = center.getY();
				double y1 = pti.getY() - yc;

				// APPLY ROTATION
				double temp_x1 = x1 * Math.cos(angle) - y1 * Math.sin(angle);
				double temp_y1 = x1 * Math.sin(angle) + y1 * Math.cos(angle);

				// TRANSLATE BACK
				arr2[i++] = plot.getWorldCoords(new ScreenPt(temp_x1 + xc, temp_y1 + yc));
			}
			_fpList.add(arr2);
		}

	}

}

File: src/firefly/js/visualize/draw/FootprintObj.js
import DrawObj from './DrawObj';
import DrawUtil from './DrawUtil';
import {makeScreenPt, makeWorldPt} from '../Point.js';
import {Style} from './DrawingDef.js';
import {startRegionDes, setRegionPropertyDes, endRegionDes} from '../region/RegionDescription.js';
import {regionPropsList, RegionType} from '../region/Region.js';
import {isEmpty} from 'lodash';
import CsysConverter from '../CsysConverter.js';

const FOOTPRINT_OBJ= 'FootprintObj';
const DEFAULT_STYLE= Style.STANDARD;
const DEF_WIDTH = 1;


/**
 * Draw one or more closed polygons.
 *
 * This class can only be used with a WebPlot, You must use WorldPt
 *
 * @author tatianag, Trey
 * @version $Id: FootprintObj.java,v 1.14 2012/11/30 23:17:01 roby Exp $
 *
 * @param {Array.<WorldPt[]>} footprintAry an array of arrays of WorldPt, each array represents 1 footprint
 * @param {Enum} style
 * @return {object}
 */
function make(footprintAry,style) {
	if (!footprintAry && !footprintAry.length) return null;

	const obj= DrawObj.makeDrawObj();
	obj.type= FOOTPRINT_OBJ;
	obj.footprintAry= footprintAry;
	if (!style) obj.style= Style.STANDARD;
	return obj;

}




////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////

const draw=  {

	usePathOptimization(drawObj) {
		return drawObj.lineWidth===1;
	},

	getCenterPt(drawObj) {
		const {footprintAry}= drawObj;
		let xSum = 0;
		let ySum = 0;
		let xTot = 0;
		let yTot = 0;

		footprintAry.forEach( (footprint) => {
			footprint.forEach( (wp) => {
				xSum += wp.x;
				ySum += wp.y;
				xTot++;
				yTot++;
			});
		});
		return makeWorldPt(xSum / xTot, ySum / yTot);
	},

	getScreenDist(drawObj,plot, pt) {
		let minDistSq = Number.MAX_VALUE;

		drawObj.footprintAry.forEach( (footprint) => {
			let totX = 0;
			let distSq;
			let totY = 0;
			let last = plot.getScreenCoords(footprint[footprint.length - 1]);
			if (!last) return 100000;
			footprint.forEach( (wpt) => {
				const testPt = plot.getScreenCoords(wpt);
				if (testPt) {
					distSq = ptSegDistSq(testPt.x, testPt.y, last.x, last.y, pt.x, pt.y);
					totX += testPt.x;
					totY += testPt.y;
					if (distSq < minDistSq) minDistSq = distSq;
					last = testPt;
				}
				else {
					totX = 100000;
					totY= 100000;
				}
			});
			const aveX = totX / footprint.length;
			const aveY = totY / footprint.length;
			distSq = distToPtSq(aveX, aveY, pt.x, pt.y);
			if (distSq < minDistSq) minDistSq = distSq;
		});

		return Math.sqrt(minDistSq);
	},

	draw(drawObj,ctx,plot,def,vpPtM,onlyAddToPath) {
		const drawParams= makeDrawParams(drawObj,def);
		if (!onlyAddToPath && drawObj.fill && drawObj.fillStyle) ctx.fillStyle= drawObj.fillStyle;
		drawFootprint(ctx, plot, drawObj.footprintAry, drawParams, drawObj.renderOptions, onlyAddToPath, drawObj.fill??false);
	},

	toRegion(drawObj, plot, def) {
		return toRegion(drawObj.footprintAry, plot, makeDrawParams(drawObj,def),drawObj.renderOptions);
	},

	translateTo(drawObj,plot, apt) {
		Object.assign({},drawObj, {footprintAry:translateTo(drawObj.footprintAry,plot,apt)});
	},

	rotateAround(drawObj, plot, angle, worldPt) {
		Object.assign({},drawObj, {footprintAry:rotateAround(drawObj.footprintAry,worldPt)});
	}
};

export default {make,draw, FOOTPRINT_OBJ};

////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////
////////////////////////////////////////////////


function makeDrawParams(fpObj,def) {
	const style= fpObj.style || def.style || DEFAULT_STYLE;
	const lineWidth= fpObj.lineWidth || def.lineWidth || DEF_WIDTH;
	return {
		color: DrawUtil.getColor(fpObj.color,def.color),
		lineWidth,
		style
	};
}





function distToPtSq(x0, y0, x1, y1) {
	const dx = x1 - x0;
	const dy = y1 - y0;
	return dx * dx + dy * dy;
}

/**
 * Returns the square of the distance from a point to a line segment. The
 * distance measured is the distance between the specified point and the
 * closest point between the specified endpoints. If the specified point
 * intersects the line segment in between the endpoints, this method returns
 * 0.0.
 *
 * COPIED from java.awt.geom.Line2D, all documentation of this method is
 * from Line2D.
 */
/**
 *
 * @param {number} x1 the coordinates of the beginning of the specified line segment
 * @param {number} y1 the coordinates of the beginning of the specified line segment
 * @param {number} x2 the coordinates of the end of the specified line segment
 * @param {number} y2 the coordinates of the end of the specified line segment
 * @param {number} px the coordinates of the specified point being measured against the specified line segment
 * @param {number} py the coordinates of the specified point being measured against the specified line segment
 * @return {number} a double value that is the square of the distance from the
 *         specified point to the specified line segment.
 */
function ptSegDistSq(x1, y1, x2, y2, px, py) {
	// Adjust vectors relative to x1,y1
	// x2,y2 becomes relative vector from x1,y1 to end of segment
	x2 -= x1;
	y2 -= y1;
	// px,py becomes relative vector from x1,y1 to test point
	px -= x1;
	py -= y1;
	let dotprod = px * x2 + py * y2;
	let projlenSq;
	if (dotprod <= 0) {
		// px,py is on the side of x1,y1 away from x2,y2
		// distance to segment is length of px,py vector
		// "length of its (clipped) projection" is now 0.0
		projlenSq = 0.0;
	} else {
		// switch to backwards vectors relative to x2,y2
		// x2,y2 are already the negative of x1,y1=>x2,y2
		// to get px,py to be the negative of px,py=>x2,y2
		// the dot product of two negated vectors is the same
		// as the dot product of the two normal vectors
		px = x2 - px;
		py = y2 - py;
		dotprod = px * x2 + py * y2;
		if (dotprod <= 0) {
			// px,py is on the side of x2,y2 away from x1,y1
			// distance to segment is length of (backwards) px,py vector
			// "length of its (clipped) projection" is now 0.0
			projlenSq = 0.0;
		} else {
			// px,py is between x1,y1 and x2,y2
			// dotprod is the length of the px,py vector
			// projected on the x2,y2=>x1,y1 vector times the
			// length of the x2,y2=>x1,y1 vector
			projlenSq = (dotprod * dotprod) / (x2 * x2 + y2 * y2);
		}
	}
	// Distance to line is now the length of the relative point
	// vector minus the length of its projection onto the line
	// (which is zero if the projection falls outside the range
	// of the line segment).
	let lenSq = px * px + py * py - projlenSq;
	if (lenSq < 0) lenSq = 0;
	return lenSq;
}


function drawFootprint(ctx, plot, footprintAry, drawParams, renderOptions, onlyAddToPath, fill=false) {
	let inView = false;
	let footprint= null;
	let pt= null;

	if (plot) {
        for (footprint of footprintAry) {
            for (pt of footprint) {
                if (pt && plot.pointOnDisplay(pt)) {
                    inView = true;
                    break;
                }
            }
            if (inView) break;
        }
	}

	if (inView || !plot) {
		for (footprint of footprintAry) {
			drawStandardFootprint(ctx, footprint, plot, drawParams, renderOptions, onlyAddToPath, fill);
		}
	}
}


function drawStandardFootprint(ctx, footprint, plot, drawParams, renderOptions, onlyAddToPath, fill= false) {

	let wpt0 = footprint[footprint.length - 1];
	let wpt= null;
	let pt0;
	let pt;
	const {color,lineWidth} = drawParams;
	let fillPath;
	if (!onlyAddToPath) {
		DrawUtil.beginPath(ctx,color,lineWidth, renderOptions);
		if (fill) {
			 fillPath= new Path2D();
		}
	}
	let firstTime= true;
	for (wpt of footprint) {
        pt0 = plot ? plot.getDeviceCoords(wpt0) : wpt0;
        pt = plot ? plot.getDeviceCoords(wpt) : wpt;
		if (!pt0 || !pt) {
            if (!onlyAddToPath) DrawUtil.stroke(ctx);
			return;
        }
		ctx.moveTo(pt0.x, pt0.y);
		if (fill && firstTime) {
			firstTime= false;
			fillPath.moveTo(pt0.x, pt0.y);
		}
		if (!plot || !plot.coordsWrap(wpt0, wpt)) {
			ctx.lineTo(pt.x, pt.y);
			if (fill) {
				fillPath.lineTo(pt0.x, pt0.y);
			}
		}
		wpt0 = wpt;
	}
	if (!onlyAddToPath) {
		if (fill ) {
			fillPath.closePath();
			DrawUtil.fill(ctx,fillPath);
		}
		else {
			DrawUtil.stroke(ctx);
		}
	}
}


function toRegion(footprintAry, plot, drawParams) {
	const {color, lineWidth} = drawParams;
	const cc = CsysConverter.make(plot);

	return footprintAry.map( (footprint) => {
				const wpAry = footprint.map((onePoint) => cc.getWorldCoords(onePoint));

				let oneDes = startRegionDes(RegionType.polygon, cc, wpAry, null, null);
				if (isEmpty(oneDes)) return '';

				oneDes += setRegionPropertyDes(regionPropsList.COLOR, color) +
					      setRegionPropertyDes(regionPropsList.LNWIDTH, lineWidth);
				oneDes = endRegionDes(oneDes);
				return oneDes;
			}).filter((oneDes) => oneDes);
}

function translateTo(footprintAry, plot, apt) {
	const pt = plot.getScreenCoords(apt);

	return footprintAry.map( (footprint) => {
		return footprint.map( (wpt) => {
			const pti = plot.getScreenCoords(wpt);
			return (pti) ? plot.getWorldCoords(makeScreenPt(pti.x + pt.x, pti.y + pt.y)) : wpt;
		});
	});
}

function rotateAround(footprintAry, plot, angle, wc) {
	const center = plot.getScreenCoords(wc);

	return footprintAry.map( (footprint) => {
		footprint.map( (p1) => DrawUtil.rotateAroundScreenPt(p1,plot,angle,center));
	});
}




